mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,281 @@
+	}
+	for group, info := range passedGroups {
+		fmt.Printf("%s: \n%+v\n", group, info)
+	}
+
+	if len(failedGroups) > 0 {
+		fmt.Println("error:")
+	}
+	for group, err := range failedGroups {
+		fmt.Printf("%s: %v\n", group, err)
+	}
+	return nil
+}
+
+func setup(conf Config) (deployer.TFPluginClient, error) {
+	network := conf.Network
+	log.Debug().Msgf("network: %s", network)
+
+	mnemonic := conf.Mnemonic
+	log.Debug().Msgf("mnemonic: %s", mnemonic)
+
+	return deployer.NewTFPluginClient(mnemonic, "sr25519", network, "", "", "", 30, false)
+}
+
+func filterNodes(tfPluginClient deployer.TFPluginClient, groups []NodesGroup, ctx context.Context) (map[string][]int, map[string]error) {
+	failedGroups := map[string]error{}
+	filteredNodes := map[string][]int{}
+	for _, group := range groups {
+		filter := types.NodeFilter{}
+		statusUp := "up"
+		filter.Status = &statusUp
+		filter.TotalCRU = &group.FreeCPU
+		filter.FreeMRU = convertMBToBytes(group.FreeMRU)
+
+		if group.FreeSRU > 0 {
+			filter.FreeSRU = convertGBToBytes(group.FreeSRU)
+		}
+		if group.FreeHRU > 0 {
+			filter.FreeHRU = convertGBToBytes(group.FreeHRU)
+		}
+		if group.Regions != "" {
+			filter.Region = &group.Regions
+		}
+		if group.Certified {
+			certified := "Certified"
+			filter.CertificationType = &certified
+		}
+		if group.Pubip4 {
+			filter.IPv4 = &group.Pubip4
+		}
+		if group.Pubip6 {
+			filter.IPv6 = &group.Pubip6
+		}
+		if group.Dedicated {
+			filter.Dedicated = &group.Dedicated
+		}
+
+		freeSSD := []uint64{group.FreeSRU}
+		if group.FreeSRU == 0 {
+			freeSSD = nil
+		}
+		freeHDD := []uint64{group.FreeHRU}
+		if group.FreeHRU == 0 {
+			freeHDD = nil
+		}
+
+		nodes, err := deployer.FilterNodes(ctx, tfPluginClient, filter, freeSSD, freeHDD, nil, int(group.NodesCount))
+		if err != nil {
+			failedGroups[group.Name] = err
+			continue
+		}
+
+		nodesIDs := []int{}
+		for _, node := range nodes {
+			nodesIDs = append(nodesIDs, node.NodeID)
+		}
+		filteredNodes[group.Name] = nodesIDs
+	}
+	return filteredNodes, failedGroups
+}
+
+func parseVMs(tfPluginClient deployer.TFPluginClient, vms []Vms, nodeGroups map[string][]int, sshKeys map[string]string) map[string]groupDeploymentsInfo {
+	deploymentsInfo := map[string]groupDeploymentsInfo{}
+	vmsOfNodeGroups := map[string][]Vms{}
+	for _, vm := range vms {
+		vmsOfNodeGroups[vm.Nodegroup] = append(vmsOfNodeGroups[vm.Nodegroup], vm)
+	}
+
+	for nodeGroup, vms := range vmsOfNodeGroups {
+		deploymentsInfo[nodeGroup] = buildDeployments(vms, nodeGroups[nodeGroup], sshKeys)
+	}
+	return deploymentsInfo
+}
+
+func massDeploy(tfPluginClient deployer.TFPluginClient, ctx context.Context, deployments groupDeploymentsInfo) ([]string, error) {
+	err := tfPluginClient.NetworkDeployer.BatchDeploy(ctx, deployments.networkDeployments)
+	if err != nil {
+		return []string{}, err
+	}
+
+	err = tfPluginClient.DeploymentDeployer.BatchDeploy(ctx, deployments.vmDeployments)
+	if err != nil {
+		return []string{}, err
+	}
```

Comment:
In case of errors, some deployments might be created here so we need to make sure before returning that anything created is destroyed.